sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -65,11 +65,66 @@ class EquivalentExpressions {
     }
   }
 
+  /**
+   * Adds only expressions which are common in each of given expressions, in a recursive way.
+   * For example, given two expressions `(a + (b + (c + 1)))` and `(d + (e + (c + 1)))`,
+   * the common expression `(c + 1)` will be added into `equivalenceMap`.
+   */
+  def addCommonExprs(exprs: Seq[Expression], addFunc: Expression => Boolean = addExpr): Unit = {
+    var exprSetForAll = ExpressionSet()
```

Comment:
Just a question; Any reason to use `ExpressionSet` instead of `mutable.Set[Expr]`? I thought we can write it like this;
```
  def addCommonExprs(exprs: Seq[Expression], addFunc: Expression => Boolean = addExpr): Unit = {
    val exprSetForAll = mutable.Set[Expr]()

    addExprTree(exprs.head, (expr: Expression) => {
      if (expr.deterministic) {
        val e = Expr(expr)
        if (exprSetForAll.contains(e)) {
          true
        } else {
          exprSetForAll += e
          false
        }
      } else {
        false
      }
    })

    val commonExprSet = exprs.tail.foldLeft(exprSetForAll) { (exprSet, expr) =>
      val otherExprSet = mutable.Set[Expr]()
      addExprTree(expr, (expr: Expression) => {
        if (expr.deterministic) {
          val e = Expr(expr)
          if (otherExprSet.contains(e)) {
            true
          } else {
            otherExprSet += e
            false
          }
        } else {
          false
        }
      })
      exprSet.intersect(otherExprSet)
    }

    commonExprSet.foreach(expr => addFunc(expr.e))
  }
```

Reply:
They actually the same effect. `Expr` compares expressions by `semanticeEquals` which compares `canonicalized` expressions. `ExpressionSet` compares `canonicalized` internally too.

Reply:
How about the overhead? It seems the current code clones a whole expr set per update though.

Reply:
Sounds good point. `ExpressionSet` has an overhead of maintaining `originals`.